### PR TITLE
Add Distributed Checkpointing support

### PR DESCRIPTION
**Pull request by ashbhandare**
This PR adds the following support:
save_checkpoint() and load_checkpoint() support to ORTTrainer.
Ability to save and load in the following scenarios:
1) Save zero checkpoints, load in zero-enabled model with same number of ranks
2) Load partitioned checkpoints(FP32 weights, FP16 weights, Optimizer params) into non-partitioned run
3) Save and load in non-zero case.

It also changes the way moment names are generated to make the naming deterministic for ease of save and load to the form:
Moment_<1 or 2>_<weight_name>_view _ <view_num>

### CODE REVIEW

File: onnxruntime/test/python/onnxruntime_test_ort_trainer.py [hunk 1]
```diff
@@ -313,5 +329,63 @@ def testBertTrainingGradientAccumulation(self):
...
+        torch.manual_seed(1)
+        onnxruntime.set_seed(1)
+        model,_,device = create_ort_trainer(gradient_accumulation_steps=1,
+                        use_mixed_precision=False,
+                        allreduce_post_accumulation=True,
+                        use_simple_model_desc=True,
+                        loss_scaler=None)
+
+        ckpt_dir = get_name("ort_ckpt")
+        model.load_checkpoint(ckpt_dir, 'bert_toy_lamb_', is_partitioned=True)
```
Comment (jessebenson) [inline onnxruntime/test/python/onnxruntime_test_ort_trainer.py:372]: Is it possible to add a unit test that creates a ZeRO checkpoint, then loads it?  Rather than committing the checkpoint files?

File: orttraining/orttraining/python/ort_trainer.py [hunk 2]
```diff
@@ -8,6 +9,7 @@
 import torch.onnx
 import onnxruntime as ort
 from distutils.version import LooseVersion
+from .checkpointing_utils import list_checkpoint_files, Combine_Zero_Checkpoint
 
 DEFAULT_OPSET_VERSION = 10
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:15]: filename is being computed in several places, which can be tricky when changes are needed. One alternative is to create file masks in one place and use it everywhere

```python3
def get_checkpoint_name(partition_optimizer, prefix,world_rank=None, world_size=None):
    SINGLE_CHECKPOINT_FILENAME='{prefix}.ort.pt'
    MULTIPLE_CHECKPOINT_FILENAME='{prefix}.ZeRO.{world_rank}.{world_size}.ort.pt'
    if partition_optimizer:
        filename=MULTIPLE_CHECKPOINT_FILENAME.format(prefix=prefix, world_rank=world_rank,     world_size=world_size)
    else:
        filename=SINGLE_CHECKPOINT_FILENAME.format(prefix=prefix)
````

File: orttraining/orttraining/python/checkpointing_utils.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+def list_checkpoint_files(checkpoint_dir, checkpoint_prefix, extension='.tar'):
+    ckpt_file_names = [f for f in os.listdir(checkpoint_dir) if f.startswith(checkpoint_prefix)]
+    ckpt_file_names = [f for f in ckpt_file_names if f.endswith(extension)]
+    ckpt_file_names = [os.path.join(checkpoint_dir, f) for f in ckpt_file_names]
+    
+    assert len(ckpt_file_names) > 0, "No checkpoint files found with prefix \"{}\" in directory {}.".format(checkpoint_prefix, checkpoint_dir)
+    return ckpt_file_names
+
+class Combine_Zero_Checkpoint():
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/checkpointing_utils.py:13]: Python classes usually dont have underscore in their names and the best practice is to inherit from `object` when no for base class (the one that doesnt have any parent class). How about `CombineZeroCheckpoint(object)`.

**Comment by thiagocrepaldi** [comment]
Current proposal suggests that `save_checkpoint` and `load_checkpoint` to belong to `ORTTrainer`. As a consequence, when `ORTModel` is re-introduced or when new trainers are available, this implementation will have to be copied, leading to redundancy.

I suggest moving this 2 APIs to onnxruntime module directly, as opposed to onnxruntime.ORTTrainer.

Some API changes would have to happen so as to receive the model as parameter, for instance, but the result would be a modular common approach for ORTTrainer and ORTModel which is similar to Pytorch approach

File: orttraining/orttraining/python/ort_trainer.py [hunk 4]
```diff
@@ -691,6 +693,73 @@ def load_state_dict(self, state_dict, strict=False):
...
+        # Values will be overwritten for matching keys among workers
+        all_checkpoint_states=dict()
+        for checkpoint_file in checkpoint_files:
+            checkpoint_state = torch.load(checkpoint_file, map_location='cpu')
+            del(checkpoint_state['model'])
+            all_checkpoint_states.update(checkpoint_state)            
+        return all_checkpoint_states
+
+    def load_checkpoint(self, checkpoint_dir, checkpoint_prefix="ORT_checkpoint", is_partitioned=False, strict=False):
+        if (not self.partition_optimizer_) and is_partitioned:
```
[the remark this answers is not in the document]
Comment (jessebenson) [reply orttraining/orttraining/python/ort_trainer.py:758]: I believe this is to support loading a ZeRO checkpoint into a single node/non-ZeRO run.

File: orttraining/orttraining/python/ort_trainer.py [hunk 5]
```diff
@@ -691,6 +693,73 @@ def load_state_dict(self, state_dict, strict=False):
...
+        # aggregate other keys in the state_dict. 
+        # Values will be overwritten for matching keys among workers
+        all_checkpoint_states=dict()
+        for checkpoint_file in checkpoint_files:
+            checkpoint_state = torch.load(checkpoint_file, map_location='cpu')
+            del(checkpoint_state['model'])
+            all_checkpoint_states.update(checkpoint_state)            
+        return all_checkpoint_states
+
+    def load_checkpoint(self, checkpoint_dir, checkpoint_prefix="ORT_checkpoint", is_partitioned=False, strict=False):
```
[the remark this answers is not in the document]
Comment (ashbhandare) [reply orttraining/orttraining/python/ort_trainer.py:757]: There can be a scenario where there are both of these in the same directory:
bert_toy_lamb.ZeROrank.1.3.ort.pt
bert_toy_lamb.ort.pt
In that case it would be unclear which checkpoint the user is referring to.

What we could do is, if the user forgets to pass the `is_partitioned` flag and no checkpoint `bert_toy_lamb.ort.pt ` exists, we could still look if zero checkpoints exist in the directory.

File: orttraining/orttraining/python/ort_trainer.py [hunk 6]
```diff
@@ -691,6 +693,73 @@ def load_state_dict(self, state_dict, strict=False):
...
+            checkpoint_name = checkpoint_prefix + "_ZeROrank_" + str(self.world_rank) + "_of_" + str(self.world_size-1) + ".tar"
+        else:
+            checkpoint_name = checkpoint_prefix + ".tar" 
+        checkpoint_file = os.path.join(checkpoint_dir, checkpoint_name)
+
+        if os.path.exists(checkpoint_file):
+            print("WARNING: {} already exists, overwriting.".format(checkpoint_file))
+
+        print("Saving checkpoint {}".format(checkpoint_file))
+        torch.save(checkpoint_state_dict, checkpoint_file)
```
[the remark this answers is not in the document]
Comment (ashbhandare) [reply orttraining/orttraining/python/ort_trainer.py:715]: The motivation for '.tar' was that pytorch recommends a checkpoint having additional info other than the model weights to be saved as '.tar' 
Source: https://pytorch.org/tutorials/beginner/saving_loading_models.html
I agree that it can be confusing, and we can save as '.ort.pt'

File: orttraining/orttraining/python/ort_trainer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/ort_trainer.py:758]: Thanks Jesse, I understand why, I was just wondering if they are really needed to accomplish in the sense of being as simple as possible for users. The more flags, the harder for users, especially beginners :)

File: orttraining/orttraining/python/ort_trainer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/ort_trainer.py:757]: The more parameters we have on the API, more scenarios we have to cover in unit tests as we have to test everything we publish. More confused the user can be, as other frameworks doesnt use such flags.

What if we do the simplest thing and clearly document it? Something like "load_checkpoint(checkpoint_dir) must point to an existing directory with a single checkpoint in it".

This way we dont have to create folders (which open space for security issues) and we dont have to care about multiple checkpoints in the same folder (the best practice would be using one checkpoint per folder, especially for zero which handles many files for a single checkpoint).

It is noble to try covering all ways users can organize their stuff, but it will double the size of source code, make mainteinance hard, require more tests and still we will never cover everything the user can do. Keeping it simple is usually the way to go :)

[the remark this answers is not in the document]
Comment (ashbhandare) [reply orttraining/orttraining/python/ort_trainer.py]: In order to not add `is_partitioned` to the API surface, enforcing that the checkpoint folder must not contain multiple checkpoint files unidentified by the prefix, unless it is a partitioned checkpoint.

File: orttraining/orttraining/python/ort_trainer.py [hunk 7]
```diff
@@ -691,6 +693,70 @@ def load_state_dict(self, state_dict, strict=False):
             session_state[name] = state_dict[name].numpy()
         self.session.load_state(session_state, strict)
 
+    def save_checkpoint(self, checkpoint_dir, checkpoint_prefix="ORT_checkpoint", checkpoint_state_dict=None):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py]: As per offline conversation, we would like to see checkpoint API being moved away from ORTTRainer, so that ORTModel.

For such, we are ok assuming that:
1) 'model' (ORTTrainer and ORTModel) has atributes world_rank, world_size and partition_optimizer
1.1) That means ORTModel needs to add a static partition_optimizer=False to match ORTTrainer
2) self and checkpoint_state_dict can be merged into a "obj" argument. To support both ORTTrainer and ORTModel,, we need:
2.1) Fort ORTTrainer, obj can be ORTTrainer or a dict which has an ORTTrainer inside. The name for this name is not really important, but can be assumed to be 'model', if preferred. Otherwise, just iterate over the dict looking for an instance of ORTTrainer and you get the name
2.2) For ORTModel, obj must be a dict with both ORTModel and torch.optim.Optimizer instances. 'model' and 'optimizer' can be assumed or discovered during dict iteration
2.3) Both models can persist additional elements in the dictionary
3) replace checkpoint_prefix by 'tag', like DeepSpeed API

File: orttraining/orttraining/python/ort_trainer.py [hunk 8]
```diff
@@ -515,6 +517,70 @@ def create_ort_training_session_bind_parameters(model, device, world_rank=-1, wo
 
     return session, train_io_binding, eval_io_binding, output_name, torch_params, output_types
 
+def save_checkpoint(model, checkpoint_dir, checkpoint_prefix="ORT_checkpoint", checkpoint_state_dict=None):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:520]: As per offline discussion, we would like to see the Checkpoint API decoupled from ORTTrainer. That makes it more portable in general, especially for the upcoming ORTModel, in which the user can pass the optimizer as part of checkpoint_state_dict

To accomplish this, we envision the following changes are necessary.

1) Merge model and checkpoint_state_dict into a single 'obj' as the first parameter of the save_checkpoint API
1.1) obj can be either ORTTrainer, ORTModel or an dictionary with them inside. For the ORTModel scenario, additionally to ORTModel instance itself, an instance of torch.optimi.Optimizer has to be present in the dictionary
1.2) The model and (for ORTModel scenario) optimizer can assume fixed (and documented) names as 'model' and 'opmizer' inside the dict or they can have arbitrary name that are discovered by iterating over the dict during input validation
1.3) Besides model and optimizer state, users can add arbitrary (key, value) entries on 'obj' for persistence
2) checkpoint_state_dict can be renamed to tag to be similar to DeepSpeed and without a default value. That helps supporting multiple checkpoints in the same folder through tag pattern matching
3) As ORTModel doesnt support ZeRO, it is OK to add a static variable called ORTModel.partition_optimizer=False to ensure that both ORTModel and ORTTrainer have mathcing APIs to support checkpoint. We understand that both models already suppoert world_rank and world_size

File: orttraining/orttraining/python/ort_trainer.py [hunk 9]
```diff
@@ -515,6 +517,70 @@ def create_ort_training_session_bind_parameters(model, device, world_rank=-1, wo
...
+    # aggregate other keys in the state_dict. 
+    # Values will be overwritten for matching keys among workers
+    all_checkpoint_states=dict()
+    for checkpoint_file in checkpoint_files:
+        checkpoint_state = torch.load(checkpoint_file, map_location='cpu')
+        del(checkpoint_state['model'])
+        all_checkpoint_states.update(checkpoint_state)            
+    return all_checkpoint_states
+
+def load_checkpoint(model, checkpoint_dir, checkpoint_prefix="ORT_checkpoint", strict=False):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/ort_trainer.py:572]: Similarly to save_checkpoint, load_checkpoint can take an "obj" as opposed to 'model'. The same conventions for find the model and optimizer can be made here.

ORTTrainer loads obj['model'] and return and ORTModel can raise and exception of NotImplementedError

checkpoint_prefix can be renamed to a mandatory tag parameter too
